lib/entities/context/RelayQuoteContext.ts

```diff
@@ -0,0 +1,143 @@
+
+  public requestKey: string;
+  public classicKey: string;
+  public routeToNativeKey: string;
+  public needsRouteToNative: boolean;
+
+  constructor(_log: Logger, public request: RelayRequest, providers: RelayQuoteContextProviders) {
+    this.log = _log.child({ context: 'RelayQuoteContext' });
+    this.rpcProvider = providers.rpcProvider;
+    this.requestKey = this.request.key();
+  }
+
+  // Relay quotes have one external dependencies:
+  // - classic request to be built from
+  dependencies(): QuoteRequest[] {
+    const classicRequest = new ClassicRequest(this.request.info, {
+      protocols: [Protocol.MIXED, Protocol.V2, Protocol.V3],
+      simulateFromAddress: this.request.config.swapper,
+      deadline: DEFAULT_ROUTING_API_DEADLINE,
+      recipient: this.request.config.swapper,
+      gasToken: this.request.config.gasToken,
+    });
+    this.classicKey = classicRequest.key();
+    this.log.info({ classicRequest: classicRequest.info }, 'Adding base classic request');
+
+    return [this.request, classicRequest];
+  }
+
+  async resolveHandler(dependencies: QuoteByKey): Promise<Quote | null> {
+    const classicQuote = dependencies[this.classicKey] as ClassicQuote;
+    const relayQuote = dependencies[this.requestKey] as RelayQuote;
+
+    const [quote] = await Promise.all([this.getRelayQuote(relayQuote, classicQuote)]);
+
+    if (!quote) {
+      this.log.warn('No Relay quote');
+      return null;
+    }
+
+    return quote;
+  }
+
+  // return either the relay quote or a constructed relay quote from  classic dependency
+  async resolve(dependencies: QuoteByKey): Promise<Quote | null> {
+    const quote = await this.resolveHandler(dependencies);
+    if (!quote || (quote as RelayQuote).amountOutEnd.eq(0)) return null;
+    return quote;
+  }
+
+  async getRelayQuote(quote?: RelayQuote, classicQuote?: ClassicQuote): Promise<RelayQuote | null> {
+    // No relay quote or classic quote
+    if (!quote && !classicQuote) return null;
+    if (!quote && classicQuote) {
+      quote = RelayQuote.fromClassicQuote(this.request, classicQuote);
+    }
+
+    if (!quote) return null;
+
+    // if its invalid for some reason, i.e. too much decay then return null
+    if (!quote.validate()) return null;
+    return quote;
+  }
+
+  // TODO: might not need, keeping for now
+  isLargeOrder(log: Logger, classicQuote: Quote): boolean {
+    // gasUseEstimateUSD on other chains seem to be unreliable
+    if (
+      classicQuote.request.info.tokenInChainId !== ChainId.MAINNET ||
+      classicQuote.request.info.tokenOutChainId !== ChainId.MAINNET
+    ) {
+      return false;
+    }
+    const quoteSizeEstimateUSD = getQuoteSizeEstimateUSD(classicQuote);
+    log.info({ quoteSize: quoteSizeEstimateUSD.toString() }, 'Quote size estimate in USD');
+    return quoteSizeEstimateUSD.gte(LARGE_TRADE_USD_THRESHOLD);
+  }
+
+  async hasApprovedPermit2(request: RelayRequest): Promise<boolean> {
```

Comment:
do we need this for relayer? shouldnt affect decision making idt since they already decided they want relayer in this case and gas is at best classic anyway